### PR TITLE
op-node: Handle duplicate outgoing supervisor events

### LINKED ISSUE #16284: op-supervisor,op-node: Exhaust L1 stalled loop
During the interop rehearsal we came across a strange stalled loop between one op-supervisor and op-node (that was driving a reth node) where it just didn't do any derivation progress and was stuck at the same L1 block. We only have info logs:

*op-supervisor:*
```
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120009-1 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-0-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120009 l2=0x3ea79e419b5337c92873c21b867e09f5f50cac541c9f63b65c87c2770339e811:39170 l1=0x6ba8958660cb1fb727db53f17d85b73874337bb9723ee482b6154d0617db72d9:8477038
t=2025-06-04T19:01:19+0000 lvl=debug msg="Next L1 block is not yet available" syncnode=syncnode-420120009-1 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-0-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120009 l1Block=0x6ba8958660cb1fb727db53f17d85b73874337bb9723ee482b6154d0617db72d9:8477038 err="not found"
t=2025-06-04T19:01:19+0000 lvl=info msg="Node completed syncing" syncnode=syncnode-420120010-2 endpoint=RPCSyncSource(ws://op-node.rehearsal-0-bn-1-opn-reth-f-snapsync-2.svc.cluster.local:9645) chain=420120010 l2=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040 l1=0xba0963e26a505b363882a50aaa2692f69ff38c6526b8425d05b0aa3e6f0d2148:8473340
```
I also included one `Node completed syncing` request by the _other_ node it was managing, which was making progress just fine.

*op-node:*
```
t=2025-06-04T19:00:16+0000 lvl=warn msg="Received signal for L1 block, but needed different block" current=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 next=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
t=2025-06-04T19:00:16+0000 lvl=info msg="Exhausted L1 data" mode=managed chainId=420120010 derivedFrom=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 derived=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040
t=2025-06-04T19:00:16+0000 lvl=info msg="Received next L1 block" mode=managed chainId=420120010 nextL1=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
t=2025-06-04T19:00:16+0000 lvl=warn msg="Received signal for L1 block, but needed different block" current=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 next=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
t=2025-06-04T19:00:16+0000 lvl=info msg="Exhausted L1 data" mode=managed chainId=420120010 derivedFrom=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 derived=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040
t=2025-06-04T19:00:16+0000 lvl=info msg="Received next L1 block" mode=managed chainId=420120010 nextL1=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
t=2025-06-04T19:00:16+0000 lvl=warn msg="Received signal for L1 block, but needed different block" current=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 next=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
t=2025-06-04T19:00:16+0000 lvl=info msg="Exhausted L1 data" mode=managed chainId=420120010 derivedFrom=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 derived=0x1ae011189eb33430aecdfb56537d8926da93d18f87437a6213f4154ac349bb65:34040
t=2025-06-04T19:00:16+0000 lvl=info msg="Received next L1 block" mode=managed chainId=420120010 nextL1=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
t=2025-06-04T19:00:16+0000 lvl=warn msg="Received signal for L1 block, but needed different block" current=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341 next=0x449328840338733e94a21b0980513cfd061e1e1a52ee673f3fe864882f27c0ed:8473341
```
In a loop stuck at the same block.

Image

A restart of the op-node fixed the problem initially, it did a reset and then synced fine again. However, after a few minutes it started becoming very slow again.

### PR DESCRIPTION
**Pull request by sebastianst**
**Description**

* Removes emitting `StepReqEvent` inside `SyncDeriver.onIncomingP2PBlock` - this was causing an avalanche of exhaust l1 events
* Deduplicate sending the same supervisor events from a managed node with a TTL until the same message is resend.
    * Complete duplication is also dangerous, as a supervisor may be restarting and miss state.

**Tests**

Added unit test of new `eventTimestamp` type and unit test of event handling behavior of `ManagedNode`.

**Additional context**

As long as we haven't fully verified that the node isn't sending a burst of duplicate events in some edge cases, it is safer to deduplicate sending events to the supervisor in this way.

**Metadata**

Fixes https://github.com/ethereum-optimism/optimism/issues/16284

### CODE REVIEW

**Review by axelKingsley — COMMENTED** [review]
Wondering out loud: would this timestamp TTL also be useful for reducing things like `UpdateCrossUnsafeEvent` during deep sync operations?

**Comment by sebastianst** [comment]
> Wondering out loud: would this timestamp TTL also be useful for reducing things like `UpdateCrossUnsafeEvent` during deep sync operations?

@axelKingsley yes we could totally leverage the new `eventTimestamp` type and accompanying pattern to deduplicate a few other events hammering the system. Good idea for a follow-up.

edit: although the cleaner fix imo is to make the code more robust and not send duplicates in the first place. I would like for us to switch some events/triggers to be range-based instead of single-block based.

File: op-node/rollup/driver/state.go [hunk 1]
```diff
@@ -280,7 +280,6 @@ func (s *SyncDeriver) onIncomingP2PBlock(envelope *eth.ExecutionPayloadEnvelope)
 	if s.SyncCfg.SyncMode == sync.CLSync || !s.Engine.IsEngineSyncing() {
 		s.Log.Info("Optimistically queueing unsafe L2 execution payload", "id", envelope.ExecutionPayload.ID())
 		s.Emitter.Emit(clsync.ReceivedUnsafePayloadEvent{Envelope: envelope})
-		s.Emitter.Emit(StepReqEvent{})
```
Comment (protolambda) [inline op-node/rollup/driver/state.go:283]: Removing this makes sense; it's a legacy thing from before we made received payloads process by the event itself and when the step-function was doing more than now.

**Review by protolambda — COMMENTED** [review]
The changes make sense, although I am still slightly concerned with the L1 traversal itself.

File: op-node/rollup/driver/state.go (hunk 1 above)
Comment (sebastianst) [reply op-node/rollup/driver/state.go:283]: Thanks for confirming

**Comment by sebastianst** [comment]
> The changes make sense, although I am still slightly concerned with the L1 traversal itself.

~My initial implementation had a bug where the derived event was deduplicated based solely on the L1 block ID, which of course doesn't make sense for the derived event, which is also emitted for every new L2 block. This event is now deduplicated based on a combined identifier of the L1 derivedFrom and L2 derived block id. Not sure if this is related to your concerns.~ Edit: I could actually revert that change. It's enough to deduplicate based on L1 ref for the L1 traversal trigger and L2 ref for the local unsafe trigger.

File: op-node/rollup/interop/managed/system.go [hunk 2]
```diff
@@ -122,55 +156,93 @@ func (m *ManagedMode) AttachEmitter(em event.Emitter) {
...
 		ref := x.Ref.BlockRef()
 		m.events.Send(&supervisortypes.ManagedEvent{UnsafeBlock: &ref})
+
 	case engine.LocalSafeUpdateEvent:
+		logger := m.log.New("derivedFrom", x.Source, "derived", x.Ref)
 		if !m.cfg.IsInterop(x.Ref.Time) {
-			m.log.Debug("Ignoring non-Interop local safe update", "derivedFrom", x.Source, "derived", x.Ref)
+			logger.Debug("Ignoring non-Interop local safe update")
 			return false
+		} else if !m.lastSafe.Update(x.Ref.ID()) {
```
Comment (sebastianst) [inline op-node/rollup/interop/managed/system.go:185]: I should actually also deduplicate this by the BlockIDPair, not only L2 ID. It works currently because we also emit the `DerivationUpdate` in the `DeriverL1StatusEvent` L1 traversal case. Edit: so it's probably ok to deduplicate only based on L2 ID because the `LocalSafeUpdateEvent` is not emitted when L1 traverses.

File: op-node/rollup/interop/managed/system.go [hunk 3]
```diff
@@ -48,7 +56,15 @@ type ManagedMode struct {
...
 
-	events *rpc.Stream[supervisortypes.ManagedEvent]
+	events managedEventStream
+
+	// outgoing event timestamp trackers
+	lastReset         eventTimestamp[struct{}]
+	lastUnsafe        eventTimestamp[eth.BlockID]
+	lastSafe          eventTimestamp[eth.BlockID]
+	lastL1Traversal   eventTimestamp[eth.BlockID]
+	lastExhaustedL1   eventTimestamp[eth.BlockIDPair]
```
Comment (sebastianst) [inline op-node/rollup/interop/managed/system.go]: I'm starting to think that it should be sufficient to deduplicate the exhausted-L1 event based on the L1 ref only. This event shouldn't be emitted until all L2 blocks are derived from an L1 source.

**Review by teddyknox — APPROVED** [review]
Stamping so that we can optimistically merge and trigger post-merge CI.

File: op-node/rollup/interop/managed/system.go (hunk 3 above)
Comment (sebastianst) [reply op-node/rollup/interop/managed/system.go]: went back to a simple BlockID

**Review by protolambda — APPROVED** [review]
